client/state/terms/reducer.js

```diff
@@ -144,6 +147,28 @@ export function items( state = {}, action ) {
 				}
 			} );
 
+		case TERMS_ADD_REQUEST:
+			return merge( {}, state, {
+				[ action.siteId ]: {
+					[ action.taxonomy ]: {
+						[ action.temporaryId ]: action.term
+					}
+				}
+			} );
+
+		case TERMS_ADD_REQUEST_SUCCESS:
+			const taxonomyTerms = merge( {}, state );
+			taxonomyTerms[ action.siteId ][ action.taxonomy ][ action.term.ID ] = action.term;
+			delete taxonomyTerms[ action.siteId ][ action.taxonomy ][ action.temporaryId ];
+
+			return taxonomyTerms;
+
+		case TERMS_ADD_REQUEST_FAILURE:
+			const terms = merge( {}, state );
+			delete terms[ action.siteId ][ action.taxonomy ][ action.temporaryId ];
```

Comment:
The "clever" side of me thinks you could create a fall-through case to handle both `_SUCCESS` and `_FAILURE` to omit this temporary term, adding the new term only if present (i.e. success).